File: volantini/scripts/processor.js
const fs = require('fs');
const path = require('path');
const fetch = require('node-fetch');

// CONFIGURAZIONE
const API_KEY = process.env.GEMINI_API_KEY;
const INPUT_DIR = path.join(__dirname, '../input');
const OUTPUT_FILE = path.join(__dirname, '../output/offerte.json');

// Lista di modelli da tentare (tutti su endpoint v1beta)
const MODELS = [
    'gemini-1.5-flash',
    'gemini-1.5-flash-latest',
    'gemini-1.5-pro',
    'gemini-1.5-pro-latest',
    'gemini-1.0-pro-vision-latest' // Fallback vecchio ma affidabile
];

async function callGemini(modelName, base64Image) {
    // NOTA: Forziamo qui v1beta
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelName}:generateContent?key=${API_KEY}`;
    
    console.log(`📡 Tento connessione a: ${modelName}...`);

    const prompt = `Analizza questo volantino (PDF) come se fossi un estrattore dati OCR avanzato.
    Estrai un array JSON valido con le offerte.
    Struttura: [{"prodotto": "Nome Prodotto", "prezzo": 1.99, "unita": "pz"}]
    Regole:
    - Prezzi: usa il punto per i decimali (es. 1.50).
    - Ignora: indirizzi, orari, numeri telefono, slogan.
    - Se trovi scritte come "SANTAGATA" e "1.5L" vicine, uniscile in "Acqua Santagata 1.5L".
    - Rispondi SOLO con il JSON.`;

    const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            contents: [{
                parts: [
                    { text: prompt },
                    { 
                        inline_data: { 
                            mime_type: "application/pdf", 
                            data: base64Image 
                        } 
                    }
                ]
            }]
        })
    });

    if (!response.ok) {
        // Se è un errore 404, ritorna null per provare il prossimo modello
        if (response.status === 404) {
            console.log(`⚠️ Modello ${modelName} non trovato (404). Passo al prossimo.`);
            return null;
        }
        // Altri errori (es. chiave scaduta, 500)
        const errText = await response.text();
        throw new Error(`Errore HTTP ${response.status}: ${errText}`);
    }

    const data = await response.json();
    return data;
}

async function main() {
    if (!API_KEY) {
        console.error("❌ ERRORE: Manca la GEMINI_API_KEY nei Secrets!");
        process.exit(1);
    }

    if (!fs.existsSync(INPUT_DIR)) fs.mkdirSync(INPUT_DIR, { recursive: true });
    
    // Cerca PDF
    const files = fs.readdirSync(INPUT_DIR).filter(f => f.toLowerCase().endsWith('.pdf'));
    if (files.length === 0) {
        console.log("⚠️ Nessun PDF trovato.");
        return;
    }

    const pdfPath = path.join(INPUT_DIR, files[0]);
    console.log(`📄 Elaborazione file: ${files[0]}`);

    const pdfBuffer = fs.readFileSync(pdfPath);
    const base64Data = pdfBuffer.toString('base64');

    let finalJson = null;

    // 🔄 Loop tentativi modelli
    for (const model of MODELS) {
        try {
            const result = await callGemini(model, base64Data);
            if (result && result.candidates && result.candidates.length > 0) {
                finalJson = result.candidates[0].content.parts[0].text;
                console.log(`✅ Successo con il modello: ${model}`);
                break; // Uscita dal loop
            }
        } catch (err) {
            console.error(`❌ Errore con ${model}:`, err.message);
        }
    }

    if (!finalJson) {
        console.error("❌ TUTTI I TENTATIVI SONO FALLITI.");
        process.exit(1);
    }

    // Pulizia e Salvataggio
    try {
        const cleanJson = finalJson.replace(/```json/g, '').replace(/```/g, '').trim();
        // Test validità JSON
        JSON.parse(cleanJson);

        const outputDir = path.dirname(OUTPUT_FILE);
        if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });

        fs.writeFileSync(OUTPUT_FILE, cleanJson);
        console.log("💾 File offerte.json salvato correttamente!");

    } catch (e) {
        console.error("❌ L'AI ha risposto ma il JSON non è valido:", e.message);
        console.log("Raw output:", finalJson);
        process.exit(1);
    }
}

main();

File: volantini/scripts/processor.py
import os
import json
import glob
import base64
import requests

# --- CONFIGURAZIONE ---
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
INPUT_DIR = os.path.join(BASE_DIR, 'input')
OUTPUT_FILE = os.path.join(BASE_DIR, 'output', 'offerte.json')
API_KEY = os.environ.get("GEMINI_API_KEY")

# LISTA DI RISERVA: Se il primo non va, prova gli altri
MODELS = [
    "gemini-1.5-flash",
    "gemini-1.5-flash-latest",
    "gemini-1.5-pro",
    "gemini-1.5-pro-latest",
    "gemini-1.0-pro-vision-latest" # Vecchio ma solido per le immagini
]

if not API_KEY:
    print("❌ ERRORE: Manca la GEMINI_API_KEY!")
    exit(1)

def encode_pdf_to_base64(pdf_path):
    """Converte il PDF in una stringa Base64 per inviarlo via web"""
    with open(pdf_path, "rb") as f:
        return base64.b64encode(f.read()).decode('utf-8')

def call_gemini_vision(model_name, base64_data):
    """Chiama l'API REST direttamente (senza libreria Google)"""
    
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}:generateContent?key={API_KEY}"
    
    payload = {
        "contents": [{
            "parts": [
                {
                    "text": """
                    Sei un assistente AI specializzato in data entry.
                    Analizza le immagini di questo PDF (volantino offerte).
                    Estrai un array JSON con i prodotti.
                    
                    Struttura JSON richiesta:
                    [
                      {"prodotto": "Nome Marca", "prezzo": 1.99, "dettagli": "kg/pz"}
                    ]

                    Regole:
                    1. Cerca i prezzi grandi e associali al testo vicino.
                    2. Unisci nome e marca.
                    3. Ignora indirizzi e orari.
                    4. Rispondi SOLO con il JSON valido.
                    """
                },
                {
                    "inline_data": {
                        "mime_type": "application/pdf",
                        "data": base64_data
                    }
                }
            ]
        }]
    }

    print(f"📡 Tentativo connessione a: {model_name}...")
    try:
        response = requests.post(url, json=payload, headers={"Content-Type": "application/json"})
        
        if response.status_code == 200:
            return response.json()
        elif response.status_code == 404:
            print(f"⚠️ Modello {model_name} non trovato (404).")
            return None
        else:
            print(f"⚠️ Errore HTTP {response.status_code}: {response.text}")
            return None
            
    except Exception as e:
        print(f"❌ Errore di connessione: {e}")
        return None

def main():
    # 1. Cerca PDF
    pdf_files = glob.glob(os.path.join(INPUT_DIR, "*.pdf"))
    if not pdf_files:
        print("⚠️ Nessun PDF trovato.")
        return

    pdf_path = pdf_files[0]
    print(f"📄 File trovato: {os.path.basename(pdf_path)}")

    # 2. Prepara il file (Base64)
    try:
        base64_data = encode_pdf_to_base64(pdf_path)
        print(f"📦 PDF codificato ({len(base64_data)} bytes).")
    except Exception as e:
        print(f"❌ Errore lettura file: {e}")
        exit(1)

    final_result = None

    # 3. Loop tentativi modelli
    for model in MODELS:
        result = call_gemini_vision(model, base64_data)
        if result and 'candidates' in result:
            final_result = result
            print(f"✅ SUCCESSO con il modello: {model}!")
            break # Usciamo dal loop appena uno funziona
        else:
            print("🔄 Passo al prossimo modello...")

    # 4. Gestione Risultato
    if not final_result:
        print("❌ TUTTI I MODELLI HANNO FALLITO.")
        exit(1)

    try:
        content_text = final_result['candidates'][0]['content']['parts'][0]['text']
        clean_json = content_text.replace("```json", "").replace("```", "").strip()
        
        # Validazione
        parsed = json.loads(clean_json)
        
        os.makedirs(os.path.dirname(OUTPUT_FILE), exist_ok=True)
        with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
            json.dump(parsed, f, indent=2, ensure_ascii=False)
            
        print("💾 File offerte.json salvato correttamente.")

    except Exception as e:
        print(f"❌ Errore salvataggio JSON: {e}")
        # Stampa raw per debug
        print(content_text)
        exit(1)

if __name__ == "__main__":
    main()
